lib/app/app.rb

```diff
@@ -39,6 +39,7 @@ class ConnectorsWebApp < Sinatra::Base
     set :connector_class, ConnectorsSdk::Base::REGISTRY.connector_class(settings.http['connector'])
     set :job_store, ConnectorsAsync::JobStore.new
     set :job_runner, ConnectorsAsync::JobRunner.new({ max_threads: settings.worker['max_thread_count'] })
+    set :secret_storage, ConnectorsAsync::SecretStorage.new
```

Comment:
Instead of using Sinatra _settings_, I think I would just change `ConnectorsAsync::SecretStorage` to be a singleton class that other logic can more easily reference without having to pass around references to the secret storage everywhere.

Reply:
For me passing the service in the constructor is a way to control the number of dependencies in the module that's using it, and it comes from the Object Oriented world background, which might not apply here :) 

Does Sinatra has limitation about storage of things in settings? Intuitively, the fact that I use a feature called "settings" to have singletons of services that I use feel wrong indeed.

Reply:
I think it's a good call, but not only for secret_storage, but also for job_store and job_runner. Do you mind if we merge this PR and approach making the mentioned three classes singletons in a different PR?

Reply:
Singleton class makes sense to me for the 3 classes, and they don't have to be part of Sinatra settings. Agree that we can have it in a separate PR as it has nothing to do with the purpose of this PR

Reply:
It depends a lot on the lifecycle of the secrets and your API signatures. 

When you start a request, will the secrets need to be updated and have to be fetched dynamically? If not, I think a secret is a static option like another and there's no real need to pass around the secret storage class instance, you can grab the value *before* and inject it in your params.

About OOP ~ I think it's not really about the case being a singleton, but rather about deciding if connectors are smart enough to go and pick secrets themselves. e.g. would it make sense to fully isolate the two concerns and make sure the connector has all it needs to function when you call it? or is this more of behavior where the connector needs to make smart decisions and pick stuff in the secrets

Reply:
I personally like Singleton classes that encapsulate pure functions. Other usages of it produce bad surprises based on my experience.

So the JobStore and SecretStorage encapsulate the abstraction of real persistence - disk, database, or anything similar, but actually for now store stuff in-memory. Therefore, currently they are actually static per application run, but in future they can be init per request (which should be okay, they are lightweight abstractions without significant memory footprint).

JobRunner is also static, but it's an abstraction over some form of a async job processor - it can be same machine, different machine, whatever fits. Based on farther implementation of it, it can be switched to per-request with insignificant memory overhead.

Reply:
What I also dislike about Singleton pattern is that it makes sense to make constructors/initializers private in case of singleton for a more robust abstraction, but it makes testing of such classes more difficult and less predictable, especially doing integration testing when a class uses such singleton - we had a bunch of problems testing classes that use AppConfig in Enterprise Search - state of AppConfig was shared within the whole rspec run and tests had to be really careful setting/reverting settings of AppConfig, I find it quite a problem compared to just initializing a lightweight object every time you need to do some work with it.